Config via dotfile

dnlserrano opened the pull request:
Tentatively fixes #9 

Would you be able to review it please @devonestes? No  problem if you can't, just leaving it here for a while before I merge it or someone else comments on it.

Thanks for your suggestion! ❤️

dnlserrano committed:
Configure exavier via .exavier.exs dotfile

As suggested by Devon Estes in #9

dnlserrano committed:
Add task to generate base .exavier.exs file

> mix exavier.config

Creates .exavier.exs with a default `threshold` of 67
and a commented version of the `test_files_to_modules`
config parameter.

Based on the great work by @rrrene on his credo library:
https://github.com/rrrene/credo

As suggested by Devon Estes.

devonestes commented on lib/exavier/config.ex:
```diff
@@ -0,0 +1,25 @@
+
+  def exists?, do: config_file() |> File.exists?()
+
+  def config_file, do: @config_file
```
You'll probably want to do a recursive search for this file in any directory from the application root, maybe using something like: https://hexdocs.pm/elixir/Application.html#app_dir/1 and https://hexdocs.pm/elixir/Path.html#wildcard/2 to get there?

dnlserrano replied:
I'd prefer we keep it as a _"root directory file"_ (relative to the project path) for now (i.e., `./.exavier.exs`). Random recursive search might be a needless pain to have here, unless you have a strong argument in favor of it?

devonestes replied:
I wouldn’t say it’s a strong argument for this specifically, but you should consider how you want to handle alternative project structures (like umbrella projects, for example).

It’s also totally cool to say you don’t support those if you want!

dnlserrano replied:
I think it's fair for people right now to have to `cd` into the directory they want to run `exavier` in. If umbrella projects become an increasing concern, we can revisit. Sounds fair? I'll merge as is now.

denispeplin replied:
I have an umbrella app that is configured in the way it's not possible to just `cd` into a directory of an individual test and run it, so `exavier` simple don't work here.
Just my `+1` to umbrella support.

devonestes commented on lib/mix/tasks/config.ex:
```diff
@@ -0,0 +1,18 @@
+
+  alias Exavier.Config, as: C
+
+  @shortdoc "Generates base config file"
```
👍 Good call with this!

devonestes commented on priv/.exavier.exs:
```diff
@@ -0,0 +1,6 @@
+%{
```
If this is just a template, I would include all possible options, but have all of them commented out and include documentation about this configuration inline (or at least links to the documentation in the comments). That way it's clear that this is optional config and that it can work without any configuration.

dnlserrano replied:
Done! ✅

dnlserrano committed:
Comment out all generated configs

Add default `nil` pattern-match for when config file is all commented
out (`nil` is the resulting value from `Code.eval_string/2`).

As suggested by Devon Estes.

dnlserrano committed:
Bulletproof load .exavier.exs config file

If present, we validate that the file is OK. If not, we error out before
even starting an exavier mutation test run.

dnlserrano commented:
Added validation of config file structure (if it exists) before running mutation tests via `exavier.test`: d1f61050908a835e65d1417da96fdd68aa19c279.
